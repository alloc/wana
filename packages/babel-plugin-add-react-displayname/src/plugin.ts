import type { NodePath, PluginPass, types as t } from '@babel/core'
import { declare } from '@babel/helper-plugin-utils'

type Options = {
  callees?: string[]
}

export default declare(babel => {
  babel.assertVersion(7)

  const { types: t } = babel

  const FunctionVisitor: babel.Visitor<PluginPass> = {
    FunctionExpression(path, state) {
      setDisplayNameIfPossible(path, state)
    },
    ArrowFunctionExpression(path, state) {
      setDisplayNameIfPossible(path, state)
    },
  }

  return {
    name: '@wana/babel-plugin-add-react-displayname',
    visitor: {
      VariableDeclaration(path, state) {
        path.traverse(FunctionVisitor, state)
      },
    },
  }

  function setDisplayNameIfPossible(
    path: NodePath<t.FunctionExpression | t.ArrowFunctionExpression>,
    state: PluginPass
  ) {
    const prevStmt = findBlockChild(path)
    if (!prevStmt) return

    const displayName = inferDisplayName(path, state.opts as Options)
    if (!displayName || /^[$_]?[a-z]/.test(displayName)) return

    if (path.isArrowFunctionExpression()) {
      const { params, body, returnType, typeParameters } = path.node
      const bodyBlock = t.isExpression(body)
        ? t.blockStatement([t.returnStatement(body)])
        : body
      const functionExpr = t.functionExpression(
        t.identifier(displayName),
        params,
        bodyBlock,
        false,
        false
      )
      functionExpr.returnType = returnType
      functionExpr.typeParameters = typeParameters
      path.replaceWith(functionExpr)
    } else {
      const newStmt = t.expressionStatement(
        t.assignmentExpression(
          '=',
          t.memberExpression(
            t.identifier(displayName),
            t.identifier('displayName')
          ),
          t.stringLiteral(displayName)
        )
      )
      newStmt.trailingComments = prevStmt.node.trailingComments
      prevStmt.node.trailingComments = null
      prevStmt.insertAfter(newStmt)
    }
  }

  function findLeftValue(expr: NodePath<t.Expression>) {
    let path = expr.parentPath
    while (path) {
      if (path.isBlockParent() || path.isProperty()) {
        return
      }
      if (path.isAssignmentExpression()) {
        return path.node.left
      }
      if (path.isVariableDeclarator()) {
        return path.node.id
      }
      path = path.parentPath!
    }
  }

  function findBlockChild(path: NodePath) {
    while (path.parentPath) {
      if (path.parentPath.isProperty()) {
        return
      }
      if (path.parentPath.isBlockParent()) {
        return path
      }
      path = path.parentPath
    }
  }

  function inferDisplayName(
    funcPath: NodePath<t.FunctionExpression | t.ArrowFunctionExpression>,
    opts: Options
  ) {
    const parentNode = funcPath.parentPath.node
    if (t.isVariableDeclarator(parentNode)) {
      // Ignore plain function components, like "const Foo = () => { ... }"
      return null
    }
    if (t.isCallExpression(parentNode) && parentNode.callee == funcPath.node) {
      // Ignore IIFEs that return JSX, like "(() => { ... })()"
      return null
    }
    if (
      doesReturnJSX(funcPath.node.body) ||
      (opts.callees && opts.callees.some(callee => hasCallee(funcPath, callee)))
    ) {
      const left = findLeftValue(funcPath)
      if (t.isIdentifier(left)) {
        return left.name
      }
    }
    return null
  }

  function hasCallee(path: NodePath, callee: string) {
    path = path.parentPath!
    while (path.isCallExpression()) {
      if (t.isIdentifier(path.node.callee, { name: callee })) {
        return true
      }
      path = path.parentPath
    }
    return false
  }

  function doesReturnJSX(node: t.Expression | t.BlockStatement) {
    if (!node) return false
    if (t.isJSX(node)) {
      return true
    }

    if (t.isBlockStatement(node)) {
      const len = node.body.length
      if (len) {
        const lastNode = node.body[len - 1]
        if (t.isReturnStatement(lastNode)) {
          return lastNode.argument && t.isJSX(lastNode.argument)
        }
      }
    }

    return false
  }
})
