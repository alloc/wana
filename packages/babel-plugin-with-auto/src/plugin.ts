import type { NodePath, types as t } from '@babel/core'
import addReactDisplayName from '@wana/babel-plugin-add-react-displayname'
import { declare } from '@babel/helper-plugin-utils'

export default declare((babel, _options, dirname) => {
  babel.assertVersion(7)

  const { types: t } = babel

  const displayNamePlugin = addReactDisplayName(
    babel,
    { callees: ['withAuto'] },
    dirname
  )

  const wrapRender = babel.template.statements(wrapRenderTmpl, {
    plugins: ['jsx', 'typescript'],
  })

  return {
    name: '@wana/babel-plugin-with-auto',
    visitor: {
      ...displayNamePlugin.visitor,
      Program(program) {
        const importDecl = getImportDeclaration(program, 'wana')
        if (!importDecl) {
          return
        }

        program.traverse({
          CallExpression(call) {
            const callee = call.get('callee')
            if (!callee.isIdentifier({ name: 'withAuto' })) {
              return
            }

            const [renderFn] = call.get('arguments')
            if (isFunctionExpression(renderFn)) {
              // Turn "withAuto" into "withAuto.dev"
              callee.replaceWith(
                t.memberExpression(
                  t.cloneNode(callee.node),
                  t.identifier('dev')
                )
              )

              replaceReturns(renderFn)
              insertAutoRender(renderFn)
              unwrapHiddenProps(renderFn, ['$auto', '$useCommit'])
            }
          },
        })
      },
    },
  }

  function isFunctionExpression(
    arg: any
  ): arg is NodePath<t.ArrowFunctionExpression | t.FunctionExpression> {
    return (
      arg && (arg.isFunctionExpression() || arg.isArrowFunctionExpression())
    )
  }

  // Replace return statements with $result assignment.
  function replaceReturns(
    renderFn: NodePath<t.ArrowFunctionExpression | t.FunctionExpression>
  ) {
    const bodyPath = renderFn.get('body')
    if (bodyPath.isExpression()) {
      return bodyPath.replaceWith(
        t.blockStatement([
          t.expressionStatement(
            t.assignmentExpression(
              '=',
              t.identifier('$result'),
              t.cloneNode(bodyPath.node)
            )
          ),
        ])
      )
    }
    renderFn.traverse({
      Function(fn) {
        // Skip nested functions.
        fn.skip()
      },
      ReturnStatement(stmt) {
        stmt.insertAfter(t.breakStatement(t.identifier('$render')))
        if (stmt.node.argument)
          stmt.replaceWith(
            t.expressionStatement(
              t.assignmentExpression(
                '=',
                t.identifier('$result'),
                t.cloneNode(stmt.node.argument)
              )
            )
          )
      },
    })
  }

  // Wrap the render function body with auto-rendering logic.
  function insertAutoRender(
    renderFn: NodePath<t.ArrowFunctionExpression | t.FunctionExpression>
  ) {
    const bodyPath = renderFn.get('body')

    const body = t.cloneNode(bodyPath.node)
    const render: t.BlockStatement = t.isExpression(body)
      ? t.blockStatement([t.expressionStatement(body)])
      : body

    bodyPath.replaceWith(t.blockStatement(wrapRender({ render })))
  }

  function unwrapHiddenProps(
    fn: NodePath<t.ArrowFunctionExpression | t.FunctionExpression>,
    hiddenPropNames: string[]
  ) {
    const hiddenProps = hiddenPropNames.map(prop => {
      const key = t.identifier(prop)
      return t.objectProperty(key, key, false, true)
    })
    const paramPaths = fn.get('params')
    const { params } = fn.node
    if (!params.length) {
      params.push(t.objectPattern(hiddenProps))
    } else {
      const [propsParam] = paramPaths
      if (propsParam.isIdentifier()) {
        propsParam.replaceWith(
          t.objectPattern([
            ...hiddenProps,
            t.restElement(t.cloneNode(propsParam.node)),
          ])
        )
      } else {
        propsParam.assertObjectPattern()
        const props = (propsParam.node as t.ObjectPattern).properties
        props.unshift(...hiddenProps)
      }
    }
  }
})

function getImportDeclaration(
  program: NodePath<t.Program>,
  moduleName: string
) {
  return program
    .get('body')
    .find(
      stmt => stmt.isImportDeclaration() && stmt.node.source.value == moduleName
    ) as NodePath<t.ImportDeclaration> | undefined
}

const wrapRenderTmpl = `
let $result, $commit = $useCommit($auto.start())
$render: try %%render%%
finally {
  $auto.stop()
}
return (
  <>
    {$result}
    {$commit()}
  </>
)
`
