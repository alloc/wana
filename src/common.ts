export const emptyArray: readonly any[] = Object.freeze([])

export const setHidden = (obj: any, key: any, value: any) =>
  Object.defineProperty(obj, key, { value, writable: true, configurable: true })

export const isObject = (value: unknown): value is object =>
  value && typeof value == 'object'

export const isFunction = (value: unknown): value is Function =>
  typeof value == 'function'

export function rethrowError(error: Error) {
  throw error
}

export const hasOwn = Function.call.bind({}.hasOwnProperty) as (
  obj: object,
  key: keyof any
) => boolean

export const noop = () => {}

export const nope = () => false

export const todo: any = () => {
  throw Error('Not yet implemented')
}
