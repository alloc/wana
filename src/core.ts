// Observation
export { o } from './o'
export { auto, Auto } from './auto'
export { when } from './when'
export { watch, Watcher } from './watch'
export { Observer, Change } from './observable'

// Escape hatches
export { no } from './no'
export { noto } from './noto'

// Internals
export { $O, $$ } from './symbols'
