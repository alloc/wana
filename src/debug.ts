import { global } from './global'

let nextDebugId = 1

const $D = Symbol.for('wana:debug')

interface DebugState {
  name: string
  actions?: any[]
  renders?: number
}

/** Get the `Auto` object for the current `withAuto` component being rendered. */
export function getCurrentAuto() {
  return global.auto
}

/** Unsafely get the `DebugState` object for the given target. */
export function getDebug(target: any): DebugState {
  return target[$D]
}

export function setDebug<T>(target: T, debug: DebugState) {
  debug.name += '#' + nextDebugId++
  target[$D] = debug
  return target
}

/** Safely add an action to a `DebugState` object */
export function addDebugAction(target: any, action: any) {
  const debug = getDebug(target)
  if (debug && debug.actions) {
    debug.actions.push(action)
  }
}
