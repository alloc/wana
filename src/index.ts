import './shims'
export { o, Observer, Change } from './observable'
export { auto, Auto } from './auto'
export { watch, Watcher } from './watch'
export { withAuto, useAuto, useO } from './render'
export { untracked } from './global'
