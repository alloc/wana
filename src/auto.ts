import { isDev } from 'is-dev'
import { batch } from './batch'
import { rethrowError } from './common'
import { addDebugAction, getDebug } from './debug'
import { global } from './global'
import { Change, ObservedValue, Observer } from './observable'
import { $O } from './symbols'
import { useAutoValue } from './ui/useAutoValue'

/** Run an effect when its tracked values change. */
export function auto(effect: () => void, config?: AutoConfig) {
  const auto = new Auto(config || {})
  auto.run(effect)
  return auto
}

export interface AutoConfig {
  /** When true, you must call `commit` to subscribe to observed values */
  lazy?: boolean
  /** When true, react to changes immediately. By default, changes are delayed until the next microtask loop */
  sync?: boolean
  /** By default, rerun the last effect (after any delay) */
  onDirty?: Auto['onDirty']
  /** By default, errors are rethrown */
  onError?: Auto['onError']
}

export class Auto {
  lazy: boolean
  sync: boolean
  dirty = true
  onDirty: (this: Auto) => void
  onError: (this: Auto, error: Error) => void
  lastObserver: AutoObserver | null = null
  nextObserver: AutoObserver | null = null

  constructor(config: AutoConfig = {}) {
    this.lazy = !!config.lazy
    this.sync = !!config.sync
    this.onDirty = config.onDirty || this._onDirty
    this.onError = config.onError || rethrowError
  }

  /** The nonce from the last commit */
  get nonce() {
    return this.lastObserver && this.lastObserver.nonce
  }

  run<T>(effect: () => T) {
    // The "run" implementation was extracted into a React hook
    // so that "withAuto" can avoid unnecessary indirection.
    return useAutoValue(this, effect)
  }

  /** Rerun the last effect and commit its observer */
  rerun() {
    const { lastObserver } = this
    return lastObserver && this.run(lastObserver.effect)
  }

  /**
   * @internal
   * Create an observer that tracks any observable properties that are
   * accessed before the `finish` method is called.
   *
   * The given `effect` is never called, except in `rerun` calls when
   * the observer has been committed.
   */
  start(effect: () => any) {
    if (global.observe) {
      throw Error('Nested tracking is forbidden')
    }
    const observer = new AutoObserver(effect)
    global.auto = this
    global.observe = (target, key) => {
      observer.observed.add(target[$O]!.get(key))
    }
    return observer
  }

  /**
   * @internal
   * Once tracking is done, mark our observer as "ready to commit".
   * Commit the observer immediately when `this.lazy` is false.
   */
  finish(observer: AutoObserver) {
    this.dirty = false
    this.nextObserver = observer

    global.auto = null
    global.observe = null

    // Compute the observer nonce at the end, in case the observed
    // effect mutates any of the observed values.
    observer.observed.forEach(observable => {
      observer.nonce += observable.nonce
    })

    if (!this.lazy) {
      this.commit()
    }
  }

  /**
   * @internal
   * Stop tracking since an effect threw an error.
   */
  catch(error: Error) {
    global.auto = null
    global.observe = null

    this.dirty = false
    this.onError(error)
  }

  /**
   * Commit the last observer passed to `finish`, unless an observed value
   * changed since then.
   */
  commit() {
    const observer = this.nextObserver
    if (observer) {
      this.nextObserver = null
      if (this.lastObserver) {
        this.lastObserver.dispose()
      }

      // Stop observing when changes occur between run and commit.
      if (this.lazy && observer.dirty) {
        this.lastObserver = null
        return false
      }

      // Attach to the observed values.
      observer.commit(this._onChange.bind(this))
      this.lastObserver = observer
    }
    return true
  }

  dispose() {
    const { lastObserver } = this
    if (lastObserver) {
      lastObserver.dispose()
      // Prevent delayed changes.
      this.lastObserver = null
    }
    // Prevent the next commit.
    this.nextObserver = null
  }

  protected _onChange(change: Change) {
    if (isDev && getDebug(this)) {
      addDebugAction(this, change)
    }
    if (!this.dirty) {
      this.dirty = true
      this.onDirty()
    }
  }

  protected _onDirty() {
    if (this.sync) {
      this.rerun()
    } else {
      batch.run(this)
    }
  }
}

export class AutoObserver extends Observer {
  observed = new Set<ObservedValue>()
  nonce = 0

  constructor(readonly effect: () => any) {
    super()
  }

  get dirty() {
    let nonce = 0
    this.observed.forEach(observable => {
      nonce += observable.nonce
    })
    return nonce != this.nonce
  }

  commit(onChange: (change: Change) => void) {
    this.onChange = onChange
    this.observed.forEach(observable => observable.add(this))
  }
}
