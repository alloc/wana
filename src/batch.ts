import queueMicrotask from 'queue-microtask'
import { Auto, AutoObserver } from './auto'

type Effect = () => void

/** The FIFO queue of async `Auto` updates */
let runQueue: RunConfig[] = []
type RunConfig = { auto: Auto; observer: AutoObserver }

/** The queue of `withAuto` updates. Sorted by deepest child last */
const renderQueue: RenderConfig[] = []
type RenderConfig = { depth: number; effect: Effect }

export interface Batch {
  /**
   * The "run" step only works with async `Auto` objects.
   * The `rerun` method is called on the next microtask.
   *
   * While running, derived `Auto` objects may be added to the batch.
   * The "run" step lasts until all derived `Auto` objects are updated,
   * or the run count exceeds 100k.
   */
  run(auto: Auto): void
  /**
   * The "render" step is sorted by `withAuto` depth.
   * After the "run" step, all `Auto` objects are updated, and `withAuto`
   * components can render now.
   */
  render(depth: number, effect: Effect): void
}

export const batch: Batch = {
  run(auto) {
    runQueue.push({ auto, observer: auto.lastObserver! })
    flush()
  },
  render(depth, effect) {
    let i = renderQueue.findIndex(other => other.depth > depth)
    if (i < 0) i = renderQueue.length
    renderQueue.splice(i, 0, { depth, effect })
    flush()
  },
}

/** Equals true when `queueMicrotask` has been called */
let flushing = false

function flush() {
  if (!flushing) {
    flushing = true
    queueMicrotask(() => {
      let runs = 0
      for (const { auto, observer } of runQueue) {
        if (++runs > 1e5) break // Break infinite loops.
        if (auto.lastObserver == observer) {
          auto.rerun()
        }
      }
      runQueue = runQueue.slice(runs)
      renderQueue.forEach(({ effect }) => effect())
      renderQueue.length = 0
      flushing = false
      if (runQueue.length) {
        flush()
      }
    })
  }
}
