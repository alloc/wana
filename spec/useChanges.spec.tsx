import { render } from '@testing-library/react'
import React from 'react'
import { $O, Change, o, useChanges } from '../src'

describe('useChanges', () => {
  it('lets you observe Change events for an observable object', () => {
    const spy = jest.fn()
    const state = o({ count: 0 })
    const Test = () => {
      useChanges(state, spy)
      return null
    }

    render(<Test />)

    state.count++ // Replace
    delete state.count // Remove
    state.count = 0 // Add

    expect(spy.mock.calls).toMatchSnapshot()
  })

  it('uses the callback from the last render by default', () => {
    const state = o({ count: 0 })
    const Test = (props: { onChange: (change: Change) => void }) => {
      useChanges(state, props.onChange)
      return null
    }

    const spy1 = jest.fn()
    const elem = render(<Test onChange={spy1} />)

    const spy2 = jest.fn()
    elem.rerender(<Test onChange={spy2} />)

    state.count++
    expect(spy1).not.toHaveBeenCalled()
    expect(spy2).toHaveBeenCalled()
  })

  it('has a deps argument for conditionally replacing the callback', () => {
    const state = o({ count: 0 })
    const Test = (props: { onChange: (change: Change) => void }) => {
      useChanges(state, props.onChange, [])
      return null
    }

    const spy1 = jest.fn()
    const elem = render(<Test onChange={spy1} />)

    const spy2 = jest.fn()
    elem.rerender(<Test onChange={spy2} />)

    state.count++
    expect(spy1).toHaveBeenCalled()
    expect(spy2).not.toHaveBeenCalled()
  })

  it('stops observing on dismount', () => {
    const state = o({ count: 0 })
    const Test = () => {
      useChanges(state, () => {})
      return null
    }

    expect(state[$O].get($O).size).toBe(0)
    const elem = render(<Test />)
    expect(state[$O].get($O).size).toBe(1)
    elem.unmount()
    expect(state[$O].get($O).size).toBe(0)
  })

  it.todo('works with arrays')

  it.todo('works with Set objects')

  it.todo('works with Map objects')
})
