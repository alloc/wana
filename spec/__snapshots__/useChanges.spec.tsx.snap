// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useChanges lets you observe Change events for an observable object 1`] = `
Array [
  Array [
    Object {
      "key": "count",
      "oldValue": 0,
      "op": "replace",
      "target": Object {
        "count": 0,
      },
      "value": 1,
    },
  ],
  Array [
    Object {
      "key": "count",
      "oldValue": 1,
      "op": "remove",
      "target": Object {
        "count": 0,
      },
    },
  ],
  Array [
    Object {
      "key": "count",
      "op": "add",
      "target": Object {
        "count": 0,
      },
      "value": 0,
    },
  ],
]
`;
